src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseInviteService.cs

```diff
@@ -0,0 +1,227 @@
+            {
+                return emailResult;
+            }
+
+            await _userInviteRepository.Create(
+                email: email,
+                role: Role.Analyst,
+                createdById: _userService.GetUserId());
+            await _userReleaseInviteRepository.CreateManyIfNotExists(
+                releaseIds: releaseIds,
+                email: email,
+                releaseRole: Contributor,
+                emailSent: true,
+                createdById: _userService.GetUserId());
+
+            return Unit.Instance;
+        }
+
+        private async Task<Either<ActionResult, Unit>> CreateExistingUserContributorInvite(List<Guid> releaseIds,
+            Guid userId, string email, string publicationTitle)
+        {
+            // check the user doesn't already have the user release roles
+            var existingReleaseRoleReleaseIds = _contentDbContext.UserReleaseRoles
+                .AsQueryable()
+                .Where(urr =>
+                    releaseIds.Contains(urr.ReleaseId)
+                    && urr.Role == Contributor
+                    && urr.UserId == userId)
+                .Select(urr => urr.ReleaseId)
+                .ToList();
+
+            var missingReleaseRoleReleaseIds = releaseIds
+                .Where(releaseId => !existingReleaseRoleReleaseIds.Contains(releaseId))
+                .ToList();
+
+            if (missingReleaseRoleReleaseIds.IsNullOrEmpty())
+            {
+                return ValidationActionResult(UserAlreadyHasReleaseRoles);
+            }
+
+            var emailResult = await SendContributorInviteEmail(
+                publicationTitle: publicationTitle,
+                releaseIds: missingReleaseRoleReleaseIds,
+                email: email);
+            if (emailResult.IsLeft)
+            {
+                return emailResult;
+            }
+
+            await _userReleaseRoleRepository.CreateManyIfNotExists(
+                userId: userId,
+                releaseIds: missingReleaseRoleReleaseIds,
+                role: Contributor,
+                createdById: _userService.GetUserId()
+            );
+
+            await _userReleaseInviteRepository.CreateManyIfNotExists(
+                releaseIds: missingReleaseRoleReleaseIds,
+                email: email,
+                releaseRole: Contributor,
+                emailSent: true,
+                createdById: _userService.GetUserId(),
+                accepted: true);
+
+            return Unit.Instance;
+        }
+
+        private async Task<Either<ActionResult, Publication>> ValidateReleaseIds(Publication publication,
+            List<Guid> releaseIds)
+        {
+            var distinctReleaseIds = releaseIds.Distinct().ToList();
+            if (distinctReleaseIds.Count != releaseIds.Count)
+            {
+                throw new ArgumentException($"{nameof(releaseIds)} should not contain duplicates",
+                    nameof(releaseIds));
+            }
```

Comment:
Not saying we need to change this, but throwing an error here is a little unnecessary. 

As we can filter out any duplicates ourselves, there isn't really any harm if they provide duplicates and no further action needs to be taken by the function consumer.

Throwing exceptions should be reserved more for cases where the consumer has done something that would actually cause an issue e.g. another exception being thrown (null pointers, etc), the creation of invalid state, etc. This is a case where coding defensively is just making life harder for the consumer :wink:

Reply:
When I realised this was a potential issue, I did at first just filter duplicates out. 

Ben raised a comment about this in this PR and that's why it is now as it is. Now the change is made, I think I probably prefer this, as I think I'd want it to blow up if something is sending duplicates releaseIds. But at the same time, it seems like a slim possibility that duplicate releaseIds will reveal a bug elsewhere in our code.

I'll just leave this as is. You'll have to argue with Ben if you want it changed :)

Reply:
I might be confused, as it's not easy navigating all the comments, but I think I raised the point about hoisting the `distinct` higher up in the method where you originally had it so that duplicates weren't passed into this method which checks they all belong to the same publication.

I don't think it was my suggestion to raise duplicates as an exception although later on when I saw it I did leave a comment to say the `"releaseIds"` could be `nameOf(releaseIds)` to assist with any renaming as part of refactoring in future.

![image](https://user-images.githubusercontent.com/4147126/147159581-6ba688a8-13e9-45df-a388-7b7a3fe06f3e.png)

I think you might be confusing this with the comment I left about `releaseIds` being empty when building the list of release titles before sending the email. This isn't client facing and something should be caught earlier before we reach the point of sending an email to invite for an empty list of Releases. For that reason I had an issue with sending the email containing a string like 'There are no releases'. I think I would either do nothing or throw the exception as I mentioned here which is what you ended up doing. 🙂 

![image](https://user-images.githubusercontent.com/4147126/147160304-32a677aa-0657-41a6-aa75-8aef01b15ac3.png)

Reply:
Ah - my bad. For some reason I thought there was only one ArgumentException, but there are two. Getting a bit lost in all these comments. I'll be more careful before attributing changes to others in the future.

Reply:
No problem! I find the interface a little bit difficult when it comes to navigating through all the comments as well. 🙂